packages/model-viewer/src/features/scene-graph/image.ts

```diff
@@ -52,37 +51,64 @@ export class Image extends ThreeDOMElement implements ImageInterface {
   }
 
   get name(): string {
-    return (this[$sourceObject] as any).name || '';
+    return (this[$sourceObject] as GLTFImage).name || '';
   }
 
   get uri(): string|undefined {
-    return this[$uri];
+    return (this[$sourceObject] as GLTFImage).uri;
+  }
+
+  get bufferView(): number|undefined {
+    return (this[$sourceObject] as GLTFImage).bufferView;
   }
 
   get type(): 'embedded'|'external' {
     return this.uri != null ? 'external' : 'embedded';
   }
 
   async setURI(uri: string): Promise<void> {
-    this[$uri] = uri;
+    (this[$sourceObject] as GLTFImage).uri = uri;
 
     const image = await new Promise((resolve, reject) => {
       loader.load(uri, resolve, undefined, reject);
     });
 
-    const texture = this[$textureInfo][$threeTexture];
-    if (texture) {
-      // If the URI is set to null but the Image had an associated buffer view
-      // (this would happen if it started out as embedded), then fall back to
-      // the cached object URL created by GLTFLoader:
-      if (image == null &&
-          (this[$sourceObject] as GLTFEmbeddedImage).bufferView != null) {
-        texture.image = this[$bufferViewImages].get(texture);
-      } else {
-        texture.image = image;
-      }
-      texture.needsUpdate = true;
-      this[$onUpdate]();
-    }
+    const texture = this[$textureInfo][$threeTexture]!;
+    texture.image = image;
+    texture.needsUpdate = true;
+    this[$onUpdate]();
+  }
+
+  async createThumbnail(width: number, height: number): Promise<string> {
+    const scene = new Scene();
+    quadMaterial.map = this[$textureInfo][$threeTexture];
+    const mesh = new Mesh(quad, quadMaterial);
+    scene.add(mesh);
+    const camera = new OrthographicCamera(-1, 1, 1, -1, 0, 1);
+
+    const {threeRenderer} = Renderer.singleton;
+    const renderTarget = new WebGLRenderTarget(width, height);
+    threeRenderer.setRenderTarget(renderTarget);
+    threeRenderer.render(scene, camera);
+    threeRenderer.setRenderTarget(null);
+
+    const buffer = new Uint8Array(width * height * 4);
+    threeRenderer.readRenderTargetPixels(
```

Comment:
Can you not just read pixels from the source buffer with glReadPixels(). It's been a while since I've done this in GL but in Vulkan you can take a texture in video memory and transition it to an RGBA uncompressed format and write it out, I would think readPixels() basically does that.

Reply:
Honestly `canvas.toBlob()` is basically just a wrapper around `readPixels` anyway. The convenience it affords is that it'll resample to a new resolution for me automatically. If this were something that was on the performance critical path, I'd look into it more, but meh.

Reply:
Yeah I guess I wasn't thinking about the fact that this is for thumbnails, this is the best way to take advantage of the hardware linear interpolation to quickly downsample.